src/ucp/core/ucp_request.inl

```diff
@@ -493,7 +493,15 @@ ucp_request_recv_buffer_reg(ucp_request_t *req, ucp_md_map_t md_map,
 
 static UCS_F_ALWAYS_INLINE void ucp_request_send_buffer_dereg(ucp_request_t *req)
 {
-    ucp_request_memory_dereg(req->send.ep->worker->context, req->send.datatype,
+    ucp_worker_h worker;
+
+    if (ucs_unlikely(req->flags & UCP_REQUEST_FLAG_CANCELED)) {
+        worker = req->send.worker;
+    } else {
+        worker = req->send.ep->worker;
+    }
```

Comment:
if it's needed for io_demo, then it should work without ucp_request_cancel at all for active message case.